addons/mail/models/mail_mail.py

```diff
@@ -58,6 +59,41 @@ class MailMail(models.Model):
         help="Failure reason. This is usually the exception thrown by the email server, stored to ease the debugging of mailing issues.")
     scheduled_date = fields.Char('Scheduled Send Date',
         help="If set, the queue manager will send the email after the date. If not set, the email will be send as soon as possible.")
+    can_read_content = fields.Boolean(compute="_compute_can_read_content")
+    body_html_access = fields.Text(
+        'Rich-text Mail Contents',
+        help="Rich-text/HTML message",
+        compute="_compute_can_read_content",
+        inverse="_inverse_body_html_access")
+    attachment_access_ids = fields.Many2many(
+        'ir.attachment',
+        string='Mail Attachments',
+        compute='_compute_can_read_content',
+        inverse='_inverse_attachment_access_ids',
+        help='Attachments are linked to a document through model / res_id and to the message '
+             'through this field.')
+
+    @api.depends_context('uid')
+    @api.depends('mail_message_id')
+    def _compute_can_read_content(self):
+        for mail in self:
+            try:
+                mail.mail_message_id.check_access_rule('read')
+                mail.can_read_content = True
+                mail.attachment_access_ids = mail.attachment_ids
+                mail.body_html_access = mail.body_html
+            except AccessError:
+                mail.can_read_content = False
+                mail.attachment_access_ids = False
+                mail.body_html_access = False
+
+    def _inverse_body_html_access(self):
+        for mail in self:
```

Comment:
Shouldn't we allow inverse update only if can_read_content ? Otherwise anyway this will fail (or should fail).

Especially if you have no right to read -> both_html_access is False -> you may try to write it. Even if interface does not allow it, it could happen.